witness/src/Dscp/Snowdrop/Validators.hs

```diff
@@ -92,51 +74,24 @@ _baseValidator =
 -- Block configuration
 ----------------------------------------------------------------------------
 
-blkStateConfig ::
-       HasWitnessConfig
-    => SD.BlkStateConfiguration SHeader SPayload SUndo HeaderHash
-                                (SD.ERwComp Exceptions Ids Values (IOCtx chgAccum) chgAccum)
-blkStateConfig = simpleBlkStateConfiguration simpleBlkConfiguration validator
-
--- | Same as inmemoryBlkStateConfiguration, but works for our patched
--- SPayload. DSCP-175
-simpleBlkStateConfiguration ::
-       SD.BlkConfiguration SHeader SPayload HeaderHash
-    -> SD.Validator Exceptions Ids Proofs Values (IOCtx chgAccum)
-    -> SD.BlkStateConfiguration SHeader SPayload SUndo HeaderHash
+blkStateConfig
+    :: HasWitnessConfig
+    => SD.BlkStateConfiguration SHeader SPayload BlockBody (SD.Undo Ids Values) HeaderHash
                                 (SD.ERwComp Exceptions Ids Values (IOCtx chgAccum) chgAccum)
-simpleBlkStateConfiguration cfg vld =
-    SD.BlkStateConfiguration {
-      bscConfig = cfg
-    , bscApplyPayload = \(SPayload txs _) -> do
-          chg <- either SD.throwLocalError pure $
-              foldM SD.mappendChangeSet def $ map SD.txBody txs
-          undo <- SD.liftERoComp $ SD.computeUndo chg
-          forM_ txs $ \tx -> do
-              SD.liftERoComp $ SD.runValidator vld tx
-              SD.modifyRwCompChgAccum (SD.txBody tx)
-          pure undo
-    , bscApplyUndo = SD.modifyRwCompChgAccum
-    , bscStoreBlund = \blund -> do
-          let blockRef = SD.bcBlockRef cfg (SD.blkHeader $ SD.buBlock blund)
-          let chg = SD.ChangeSet $ M.singleton (SD.inj $ SD.BlockRef blockRef)
-                                               (SD.New $ SD.inj blund)
-          SD.modifyRwCompChgAccum chg
-    , bscGetBlund = SD.liftERoComp . SD.queryOne . SD.BlockRef
-    , bscBlockExists = SD.liftERoComp . SD.queryOneExists . SD.BlockRef
-    , bscGetTip =
-          SD.liftERoComp (SD.queryOne SD.TipKey) >>=
-          maybe (SD.throwLocalError @(SD.BlockStateException Ids) SD.TipNotFound)
-                (pure . SD.unTipValue)
-    , bscSetTip = \newTip' -> do
-          let newTip = SD.inj $ SD.TipValue newTip'
-          let tipChg = \cons -> SD.ChangeSet $ M.singleton (SD.inj SD.TipKey) (cons newTip)
-          oldTipMb <- SD.liftERoComp $ SD.queryOne SD.TipKey
-          -- TODO check that tip corresponds to blund storage
-          case oldTipMb of
-              Nothing                            -> SD.modifyRwCompChgAccum $ tipChg SD.New
-              Just (_ :: SD.TipValue HeaderHash) -> SD.modifyRwCompChgAccum $ tipChg SD.Upd
+blkStateConfig = blkStateConfig'
+    { SD.bscApplyPayload = \(SPayload txs _) -> SD.bscApplyPayload blkStateConfig' txs
```

Comment:
I do not understand why do we apply payload here with `bcBlkVerify` turned off.

Reply:
Sorry, didn't understand the question

Reply:
I asked someone to explain this logic. Why `bscApplyPayload` uses modified `blkStateConfig`?

Reply:
Oh, I see. A very good point indeed!

@pva701 or @astiunov ?

Reply:
This part from @pva701

Reply:
@volhovm `bscApplyPayload` doesn't use `bcBlkVerify`, it can't use it even in theory, because payload is part of block, but `bcBlkVerify` verifies whole block. 

`blkStateConfig'` is needed to match type of `inmemoryBlkStateConfiguration` and `blkStateConfig`

Reply:
>because payload is part of (a) block, but bcBlkVerify verifies (the) whole block.

So why do you change it to `mempty` then?

> A is needed to match (the) type of B

I don't understand this pattern-matching game, can you please elaborate on the exact semantics of this change?

Reply:
>So why do you change it to mempty then?

because `bcBlkVerify` isn't used in `inmemoryBlkStateConfiguration` anyway, the only place where it's used it's setting `bscConfig`, but I overwrite it

>I don't understand this pattern-matching game, can you please elaborate on the exact semantics of this change?

the exact semantic is: the logic of `blkStateConfig` is the same as `inmemoryBlkStateConfiguration` and we want to abuse this fact, but `inmemoryBlkStateConfiguration` has a little bit type of payload, and I hacked all places which are affected by the payload type (`bscApplyPayload` and `bscConfig`). It can be done in this way because `cfg` isn't used in functions of `BlkStateConfiguration` (ok it's used but in only one place where we store blund, but it's ok). It's done to avoid copy-pasting of code from snowdrop.

Reply:
ok, we are going to fix it @volhovm